frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -451,146 +446,175 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
+								if err != nil {
+									return errors.Wrapf(err, "failed to parse ref %q", mutRef)
 								}
 							}
+							var img dockerspec.DockerOCIImage
+							if err := json.Unmarshal(dt, &img); err != nil {
+								return errors.Wrap(err, "failed to parse image config")
+							}
+							d.baseImg = cloneX(&img) // immutable
+							img.Created = nil
+							// if there is no explicit target platform, try to match based on image config
+							if d.platform == nil && platformOpt.implicitTarget {
+								p := autoDetectPlatform(img, *platform, platformOpt.buildPlatforms)
+								platform = &p
+							}
+							if dgst != "" {
+								ref, err = reference.WithDigest(ref, dgst)
+								if err != nil {
+									return err
+								}
+							}
+							d.stage.BaseName = ref.String()
+							if len(img.RootFS.DiffIDs) == 0 {
+								isScratch = true
+								// schema1 images can't return diffIDs so double check :(
+								for _, h := range img.History {
+									if !h.EmptyLayer {
+										isScratch = false
+										break
+									}
+								}
+							}
+							d.image = img
 						}
-						d.image = img
-					}
-					if isScratch {
-						d.state = llb.Scratch()
-					} else {
-						d.state = llb.Image(d.stage.BaseName,
-							dfCmd(d.stage.SourceCode),
-							llb.Platform(*platform),
-							opt.ImageResolveMode,
-							llb.WithCustomName(prefixCommand(d, "FROM "+d.stage.BaseName, opt.MultiPlatformRequested, platform, emptyEnvs{})),
-							location(opt.SourceMap, d.stage.Location),
-						)
-						if reachable {
-							validateBaseImagePlatform(origName, *platform, d.image.Platform, d.stage.Location, lint)
+						if isScratch {
+							d.state = llb.Scratch()
+						} else {
+							d.state = llb.Image(d.stage.BaseName,
+								dfCmd(d.stage.SourceCode),
+								llb.Platform(*platform),
+								opt.ImageResolveMode,
+								llb.WithCustomName(prefixCommand(d, "FROM "+d.stage.BaseName, opt.MultiPlatformRequested, platform, emptyEnvs{})),
+								location(opt.SourceMap, d.stage.Location),
+							)
+							if reachable {
+								validateBaseImagePlatform(origName, *platform, d.image.Platform, d.stage.Location, lint)
+							}
 						}
-					}
-					d.platform = platform
-					return nil
-				})
-			}(i, d)
+						d.platform = platform
+						return nil
+					})
+				}(i, d)
+			}
 		}
+
+		if err := eg.Wait(); err != nil {
+			return nil, err
+		}
+		return allReachable, nil
 	}
 
-	if err := eg.Wait(); err != nil {
-		return nil, err
+	var allReachable map[*dispatchState]struct{}
+	for {
+		allReachable, err = resolveReachableStages(ctx, allDispatchStates.states, target)
+		if err != nil {
+			return nil, err
+		}
+
+		// initialize onbuild triggers in case they create new dependencies
+		newDeps := false
+		for d := range allReachable {
+			d.init()
+
+			if len(d.image.Config.OnBuild) > 0 {
+				if b, err := initOnBuildTriggers(d, d.image.Config.OnBuild, allDispatchStates); err != nil {
```

Comment:
Should we have set a prefix to show this is coming from ONBUILD?

Reply:
If it has a prefix then I think it means "add the onbuild rule to current image config" (this is what is shown in legacy builder I think). Here it is actually running the command.